### ISSUE TITLE
how to use class idx 0 as no-object class?

### ISSUE THREAD

**Issue by HotaekHan** [facebookresearch/detr#356]
## ❓ How to do something using DETR

I want class idx 0 as no-object class.
which part should I fix?

Before
<pre><code>class SetCriterion(nn.Module):
    def __init__():
        empty_weight = torch.ones(self.num_classes + 1)
        empty_weight[-1] = self.eos_coef
    def loss_labels():
        target_classes_o = torch.cat([t["labels"][J] for t, (_, J) in zip(targets, indices)])
        target_classes = torch.full(src_logits.shape[:2], self.num_classes,
                                    dtype=torch.int64, device=src_logits.device)
        target_classes[idx] = target_classes_o

        loss_ce = F.cross_entropy(src_logits.transpose(1, 2), target_classes, self.empty_weight)
</code></pre>

After
<pre><code>class SetCriterion(nn.Module):
    def __init__()
        empty_weight = torch.ones(self.num_classes + 1)
        empty_weight[0] = self.eos_coef
        self.background_idx = 0
    def loss_labels()
        target_classes_o = torch.cat([t["labels"][J] for t, (_, J) in zip(targets, indices)])
        target_classes = torch.full(src_logits.shape[:2], self.background_idx,
                                    dtype=torch.int64, device=src_logits.device)
        target_classes[idx] = target_classes_o

        loss_ce = F.cross_entropy(src_logits.transpose(1, 2), target_classes, self.empty_weight)
</code></pre>

Is it right?
and is it all things should I concern?

by the way, also have another question.
does the self.empty_weight need for the imbalance between foreground and background?

thanks :)

**Comment by alcinos** [comment]
> Is it right?
> and is it all things should I concern?

That's a good start, but there are most likely other places to fix. From the top of my head there is at least the postprocessor: https://github.com/facebookresearch/detr/blob/a54b77800eb8e64e3ad0d8237789fcbf2f8350c5/models/detr.py#L275 and the cardinality error computation: https://github.com/facebookresearch/detr/blob/a54b77800eb8e64e3ad0d8237789fcbf2f8350c5/models/detr.py#L138 but maybe I'm forgetting others.

>  does the self.empty_weight need for the imbalance between foreground and background?

Yes, there is an imbalance between the number of queries that contain an object (on average 7 in Coco) and those that don't (on average 100-7=93 on Coco)

**Comment by amirhesamyazdi** [comment]
> Yes, there is an imbalance between the number of queries that contain an object (on average 7 in Coco) and those that don't (on average 100-7=93 on Coco

Regarding no-object class. I am still very confused. I have trained a model which works great when inferring images with object. But when I want to test for false alarms and give it images with no objects, it performs very poorly and detects a lot of objects anyway. so I am very confused still.

If I want to use detr for my model that uses 3 classes instead of 70 something, what should I do to improve false alarm situation? 
Is there a particular reason why COCO annotations skip some categories in the middle? There are N/As in the middle like category_id 12 does not exist at all/ and then several throughout. So even though I only have 3 classes and in any of my training/val images there can be up to 9 objects, should I set number of classes to something significantly higher than 3? Like what you have done for COCO? I notice there is not 91 classes in COCO! just 71 or somehting.
